server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -3883,6 +3893,14 @@ private Tuple<BlobStoreIndexShardSnapshots, ShardGeneration> buildBlobStoreIndex
                     generation
                 );
             } catch (NoSuchFileException noSuchFileException) {
+                final long latestGeneration = latestIndexBlobId();
+                final RepositoryData currentRepositoryData = getRepositoryData(latestGeneration);
+                if (currentRepositoryData.shardGenerations().hasShardGen(new RepositoryShardId(indexId, shardId)) == false) {
+                    // Master has concurrently mutated the shard generation. This can happen when master fails over
+                    // which is "expected". We do not need the following workaround in this case.
+                    throw noSuchFileException;
+                }
```

Comment:
I wanted to simply compare the last known repo generation against the latest generation by listing root blobs. But it seems last known repo generation is not easily available especially when this runs in the context of shard snapshotting. Since this is a rare case, it feels ok to just read the repo data directly and check whether the shard gen still exists. It's slower but we are on a slow path already more or less. Hopefully that works for you.

That said, `BlobStoreRepositoy` class has many method related to fetching and caching repository data. Maybe one (some) of them can be reliably leveraged. Please let me know. Thank you!